fabric-block-api-v1/src/main/java/net/fabricmc/fabric/api/block/v1/FabricBlock.java

```diff
@@ -0,0 +1,105 @@
+	 * for example by calling {@link World#updateListeners}.
+	 *
+	 * <p>Note: Overriding this method for a block does <strong>not</strong> change how it renders.
+	 * It's up to modded models to check for the appearance of nearby blocks and adjust accordingly.
+	 *
+	 * <h3>Implementation guidelines</h3>
+	 *
+	 * <p>This can be called on the server, where block entity data can be safely accessed,
+	 * and on the client, possibly in a meshing thread, where block entity data is not safe to access !
+	 * Here is an example of how data from a block entity can be handled safely.
+	 * The block entity needs to implement {@code RenderAttachmentBlockEntity} for this to work.
+	 * <pre>{@code @Override
+	 * public BlockState getAppearance(BlockState state, BlockRenderView renderView, BlockPos pos, Direction side, @Nullable BlockState sourceState, @Nullable BlockPos sourcePos) {
+	 *      if (renderView instanceof ServerWorld serverWorld) {
+	 *          // Server side, ok to use block entity directly!
+	 *          BlockEntity blockEntity = serverWorld.getBlockEntity(pos);
+	 *
+	 *          if (blockEntity instanceof ...) {
+	 *              // Get data from block entity
+	 *              return ...;
+	 *          }
+	 *      } else {
+	 *      	// Client side, need to use the render attachment!
+	 *          RenderAttachedBlockView attachmentView = (RenderAttachedBlockView) renderView;
+	 *          Object data = attachmentView.getBlockEntityRenderAttachment(pos);
+	 *
+	 *     	    // Check if data is not null and of the correct type, and use that to determine the appearance
+	 *     	    if (data instanceof ...) {
+	 *     	        // get appearance for side ...
+	 *     	        return ...;
+	 *            }
+	 *      }
+	 *
+	 *     	// Example of varying the appearance based on the source pos
+	 *      if (sourcePos != null) {
+	 *          // get appearance for side ...
+	 *          return ...;
+	 *      }
+	 *
+	 *      // If there is no other appearance, just return the original block state
+	 *      return state;
+	 * });
+	 * }</pre>
+	 *
+	 * @param state       state of this block, whose appearance is being queried
+	 * @param renderView  the world this block is in
+	 * @param pos         position of this block, whose appearance is being queried
+	 * @param side        the side for which the appearance is being queried
+	 * @param sourceState (optional) state of the block that is querying the appearance, or null if unknown
+	 * @param sourcePos   (optional) position of the block that is querying the appearance, or null if unknown
+	 * @return the appearance of the block on the given side; the original {@code state} can be returned if there is no better option
+	 */
+	default BlockState getAppearance(BlockState state, BlockRenderView renderView, BlockPos pos, Direction side, @Nullable BlockState sourceState, @Nullable BlockPos sourcePos) {
```

Comment:
Thoughts on adding @ApiStatus.OverrideOnly on this to encorage using the blockstate method?